internal/controller/application_controller.go

```diff
@@ -163,49 +174,142 @@ func (r *ApplicationPullServiceAccountCreator) getImageRepositoryPullSecretNames
+	})
+
+	for _, secretName := range individualSecretNames {
+		secret := &corev1.Secret{}
+		err := r.Client.Get(ctx, types.NamespacedName{Name: secretName, Namespace: application.Namespace}, secret)
+		if err != nil {
+			if errors.IsNotFound(err) {
+				log.Info("Individual pull secret not found, skipping", "secretName", secretName)
+				continue
+			}
+			log.Error(err, "failed to get individual pull secret", "secretName", secretName, l.Action, l.ActionView)
+			return nil, err
+		}
+
+		// Only process secrets of type kubernetes.io/dockerconfigjson
+		if secret.Type == corev1.SecretTypeDockerConfigJson {
+			if dockerConfigDataBytes, ok := secret.Data[corev1.DockerConfigJsonKey]; ok {
+				var dcj dockerConfigJson
+				if err := json.Unmarshal(dockerConfigDataBytes, &dcj); err != nil {
+					log.Error(err, "failed to unmarshal .dockerconfigjson data from secret", "secretName", secretName)
+					continue
+				}
+				for registry, authEntry := range dcj.Auths {
+					combinedAuths[registry] = authEntry
+				}
+			} else {
+				log.Info("Secret of type kubernetes.io/dockerconfigjson found, but missing .dockerconfigjson key", "secretName", secretName)
+			}
+		} else {
+			log.Info("Skipping secret due to unexpected type (only kubernetes.io/dockerconfigjson is processed)", "secretName", secretName, "type", secret.Type)
 		}
-	} else {
-		return nil
 	}
 
-	// Create service account for application
-	secretsReferences := []corev1.ObjectReference{}
-	imagePullSecretsReferences := []corev1.LocalObjectReference{}
-	for _, secretName := range pullSecretNames {
-		secretsReferences = append(secretsReferences, corev1.ObjectReference{Name: secretName})
-		imagePullSecretsReferences = append(imagePullSecretsReferences, corev1.LocalObjectReference{Name: secretName})
+	// Marshal combined auths back into .dockerconfigjson format
+	combinedDockerConfig := dockerConfigJson{Auths: combinedAuths}
+	marshaledData, err := json.Marshal(combinedDockerConfig)
+	if err != nil {
+		log.Error(err, "failed to marshal combined docker config json")
+		return nil, err
 	}
 
-	applicationSA := corev1.ServiceAccount{
+	// Create or update the aggregated secret
+	aggregatedSecret := &corev1.Secret{
 		ObjectMeta: metav1.ObjectMeta{
-			Name:      serviceAccountName,
+			Name:      aggregatedSecretName,
 			Namespace: application.Namespace,
 		},
-		Secrets:          secretsReferences,
-		ImagePullSecrets: imagePullSecretsReferences,
+		Type: corev1.SecretTypeDockerConfigJson,
+		Data: map[string][]byte{
+			corev1.DockerConfigJsonKey: marshaledData,
+		},
 	}
 
-	// set serviceAccount ownership to application
-	if err := controllerutil.SetOwnerReference(application, &applicationSA, r.Scheme); err != nil {
-		log.Error(err, "failed to set application as owner of SA", "applicationName", application.Name)
-		return err
+	// Set pull secret ownership to application
+	if err := controllerutil.SetOwnerReference(application, aggregatedSecret, r.Scheme); err != nil {
```

Comment:
call this just before creating, no need to set it for already existing one

Reply:
Kept only this and removed the second call.